lib/common/http1client.c

```diff
@@ -374,11 +388,112 @@ static void on_send_request(h2o_socket_t *sock, const char *err)
         return;
     }
 
+    if (client->_is_chunked) {
+        client->_is_chunked = 0;
+        h2o_iovec_t last = h2o_iovec_init(H2O_STRLIT("0\r\n"));
+        h2o_socket_write(client->super.sock, &last, 1, on_send_request);
+        return;
+    }
+
     h2o_socket_read_start(client->super.sock, on_head);
     client->_timeout.cb = on_head_timeout;
     h2o_timeout_link(client->super.ctx->loop, client->super.ctx->io_timeout, &client->_timeout);
 }
 
+static void on_req_body_done(h2o_socket_t *sock, const char *err)
+{
+    struct st_h2o_http1client_private_t *client = sock->data;
+
+    if (client->_body_buf_in_flight != NULL) {
+        client->_write_req_chunk_done.cb(client->_write_req_chunk_done.ctx, client->_body_buf_in_flight->size,
+                                         client->_body_buf_is_done);
+        h2o_buffer_consume(&client->_body_buf_in_flight, client->_body_buf_in_flight->size);
+    }
+
+    if (err) {
+        on_send_request(client->super.sock, err);
+        return;
+    }
+
+    if (client->_body_buf != NULL && client->_body_buf->size != 0)
+        h2o_http1client_write_req_chunk(client->super.sock, h2o_iovec_init(NULL, 0), client->_body_buf_is_done);
+    else if (client->_body_buf_is_done)
+        on_send_request(client->super.sock, NULL);
+}
+
+static void swap_buffers(h2o_buffer_t **a, h2o_buffer_t **b)
+{
+    h2o_buffer_t *swap;
+    swap = *b;
+    *b = *a;
+    *a = swap;
+}
+
+void write_chunk_to_socket(struct st_h2o_http1client_private_t *client, h2o_iovec_t headers_top, h2o_iovec_t headers_bottom, h2o_iovec_t chunk, h2o_socket_cb cb)
+{
+    int i = 0;
+    h2o_iovec_t chunk_and_reqbufs[5];
+
+    if (headers_top.base)
+        chunk_and_reqbufs[i++] = headers_top;
+    if (headers_bottom.base)
+        chunk_and_reqbufs[i++] = headers_bottom;
+
+
+    chunk_and_reqbufs[i].len = snprintf(client->_chunk_len_str, sizeof(client->_chunk_len_str), "%zX\r\n", chunk.len);
```

Comment:
Can we use lower-case characters for the hex digits considering the fact that we do so for the chunked response (in lib/handler/chunked.c)?